Ignore unknown enum received in json when ignoreUnknownFields flag is set

vijay-bhatt opened the pull request:
Ignore unknown enum received in json when ignoreUnknownFields flag is set.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/protobuf/4825)
<!-- Reviewable:end -->

xfxyjwf commented on java/util/src/main/java/com/google/protobuf/util/JsonFormat.java:
```diff
@@ -1746,6 +1746,10 @@ private EnumValueDescriptor parseEnum(EnumDescriptor enumDescriptor, JsonElement
           // an exception later.
+            
+          if (ignoringUnknownFields) {
+            result = enumDescriptor.getValues().get(0);
```
This will cause the field to have the default value, i.e., instead of ignoring unknown enum values, they are converted to the default value. It doesn't matter to proto3 because having the default value == not set, but for proto2 it means data corruption (users will get an incorrect enum value).

Could you find a way to ignore the field entirely instead?

vijay-bhatt replied:
I actually based this on your comment dated Jun 30 2017 for the discussion on https://github.com/google/protobuf/pull/3289/commits/70049452c2ece7918b3bf8cb992b8178e6e5e1cf
for veerenmandalia's implementation

Having to "treat it as if the field is not present" would mean assigning it a default value.

xfxyjwf replied:
"treat it as if the field is not present" does not mean assigning it a default value. In proto3, assigning default value == not present, but that's not the case in proto2. Here we need to find a way to skip the field entirely and not set any value into the field.

vijay-bhatt replied:
Ok. So instead, if i return null from this method, the protobuf field will be not be set. In my test, that is also returning the default value as the builder enum int is preset to 0 and the proto generated class returns default if the enum int value is 0. Should returning null be the fix here? How will it behave with proto2 compiler?

xfxyjwf replied:
I think the code needs some refactoring to be able to ignore unknown enum values. Returning null is one option and we need to update all callsites of the method to make sure they are handling the null return value properly.

vijay-bhatt replied:
I have made some refactoring change. I also saw the call sites where the changed method is called. In map, it is adding the null as value. As a result, it should return a default 0th enum. Accordingly I have added 2 more tests, one for repeated enum and other for enum in map value.

xfxyjwf commented on java/util/src/test/java/com/google/protobuf/util/JsonFormatTest.java:
```diff
@@ -1179,8 +1179,8 @@ public void testParserIgnoringUnknownEnums() throws Exception {
     JsonFormat.parser().ignoringUnknownFields().merge(json, builder);
-    assertEquals(NestedEnum.FOO, builder.getOptionalNestedEnum());
-  }  
+    assertEquals(0, builder.getOptionalNestedEnumValue());
```
Add a test for unknown enum value in a repeated enum field?

xfxyjwf commented on java/util/src/test/java/com/google/protobuf/util/JsonFormatTest.java:
```diff
@@ -1173,7 +1173,41 @@ public void testParserIgnoringUnknownFields() throws Exception {
+          + "}",
+        builder);
 
+    assertEquals(Integer.valueOf(0), builder.getInt32ToEnumMapValueMap().get(1));
```
The entry should be ignored, i.e., no entry for 1 should exist in the map.

xfxyjwf commented on java/util/src/test/java/com/google/protobuf/util/JsonFormatTest.java:
```diff
@@ -1173,7 +1173,41 @@ public void testParserIgnoringUnknownFields() throws Exception {
+          + "}",
+        builder);
+
+        assertEquals(0, builder.getRepeatedNestedEnumValue(0));
```
Same here. I think the entry should be discarded instead of converted to a different value.

vijay-bhatt commented:
Please review. I have refactored the code so that unknown enum in a Map of enum value is ignored. Similarly for repeated unknown enum, it also is ignored. Changed the test cases accordingly.

xfxyjwf commented on java/util/src/main/java/com/google/protobuf/util/JsonFormat.java:
```diff
@@ -1557,8 +1561,12 @@ private void mergeRepeatedField(
         if (value == null) {
-          throw new InvalidProtocolBufferException(
-              "Repeated field elements cannot be null in field: " + field.getFullName());
+          if(field.getFile().getSyntax() == FileDescriptor.Syntax.PROTO3) {
```
Same here.

xfxyjwf commented on java/util/src/test/java/com/google/protobuf/util/JsonFormatTest.java:
```diff
@@ -659,7 +659,9 @@ public void testMapNullValueIsRejected() throws Exception {
     }
+  }
 
+  public void testMapNullValueIsIgnored() throws Exception {
```
We shouldn't introduce this behavior change for non-enum fields.

vijay-bhatt commented:
Please review

xfxyjwf commented on java/util/src/main/java/com/google/protobuf/util/JsonFormat.java:
```diff
@@ -1539,7 +1540,11 @@ private void mergeMapField(FieldDescriptor field, JsonElement json, Message.Buil
         Object value = parseFieldValue(valueField, entry.getValue(), entryBuilder);
         if (value == null) {
-          throw new InvalidProtocolBufferException("Map value cannot be null.");
+          if(valueField.getType() == Type.ENUM && field.getFile().getSyntax() == FileDescriptor.Syntax.PROTO3) {
```
Can you change this to:
```
if (ignoringUnknownFields && valueField.getType() == Type.ENUM) {
  continue;
} else {
  throw ...
}
```

xfxyjwf commented on java/util/src/main/java/com/google/protobuf/util/JsonFormat.java:
```diff
@@ -1557,8 +1562,12 @@ private void mergeRepeatedField(
         if (value == null) {
-          throw new InvalidProtocolBufferException(
-              "Repeated field elements cannot be null in field: " + field.getFullName());
+          if(field.getType() == Type.ENUM && field.getFile().getSyntax() == FileDescriptor.Syntax.PROTO3) {
```
Change this to:
```
if (ignoringUnknownFields && field.getType() == Type.ENUM) {
} else {
}
```

xfxyjwf commented on java/util/src/test/java/com/google/protobuf/util/JsonFormatTest.java:
```diff
@@ -659,7 +659,9 @@ public void testMapNullValueIsRejected() throws Exception {
     }
+  }
 
+  public void testMapEnumNullValueIsIgnored() throws Exception {
```
Please keep the old "testMapNullValueIsRejected()" method as-is, and add a new "testMapEnumNullValueIsIgnored()" method.

xfxyjwf commented on java/util/src/test/java/com/google/protobuf/util/JsonFormatTest.java:
```diff
@@ -668,7 +670,9 @@ public void testMapNullValueIsRejected() throws Exception {
               + "}",
           builder);
-      fail();
+      assertEquals(0, builder.getInt32ToEnumMapMap().size());
```
Please don't change this test for non-enum map values. As I suggested above, add a new method instead (do not split the code from the existing "testMapNullValueIsRejected()" method.

vijay-bhatt commented:
I have incorporated all your code review comments. Please review.

xfxyjwf commented on java/util/src/test/java/com/google/protobuf/util/JsonFormatTest.java:
```diff
@@ -30,6 +30,18 @@
 
 package com.google.protobuf.util;
 
+import java.io.IOException;
```
Please restore the original imports order...

vijay-bhatt commented:
Done

vijay-bhatt commented:
Will these changes be pulled or do I need to merge / push?

xfxyjwf commented:
I just started the tests and will merge after the tests finish.

vijay-bhatt commented:
Thanks
